topic_table_model.cpp

```diff
@@ -531,15 +531,187 @@ CORBA::TCKind TopicTableModel::typekind_to_tckind(DDS::TypeKind tk)
+    }
+    case CORBA::tk_ulonglong: {
+        CORBA::ULongLong value;
+        if (check_rc(data->get_uint64_value(value, id), "get_uint64_value failed")) {
+            data_row->value = static_cast<quint64>(value);
+        }
+        break;
+    }
+    case CORBA::tk_string: {
+        CORBA::String_var value;
+        if (check_rc(data->get_string_value(value, id), "get_string_value failed")) {
+            data_row->value = value.in();
+        }
+        break;
+    }
+    case CORBA::tk_enum: {
+        // When @bit_bound is supported, update this to call the right interface.
+        CORBA::Long value;
+        if (check_rc(data->get_int32_value(value, id), "get enum value failed")) {
+            data_row->value = static_cast<int32_t>(value);
+        }
+        break;
+    }
+    default:
+        data_row->value = "NULL";
+        break;
+    }
 }
 
 //------------------------------------------------------------------------------
-void TopicTableModel::parseData(const DDS::DynamicData_var& data)
+void TopicTableModel::parseCollection(const DDS::DynamicData_var& data, std::string namePrefix)
+{
+    DDS::DynamicType_var type = data->type();
+    DDS::TypeDescriptor_var td;
+    if (type->get_descriptor(td) != DDS::RETCODE_OK) {
+        std::cerr << "Failed to get TypeDescriptor for a "
+                  << OpenDDS::XTypes::typekind_to_string(type->get_kind())
+                  << " member" << std::endl;
+        return;
+    }
+    const unsigned int count = data->get_item_count();
+    DDS::DynamicType_var elem_type = OpenDDS::XTypes::get_base_type(td->element_type());
+    const OpenDDS::XTypes::TypeKind elem_tk = elem_type->get_kind();
+
+    for (unsigned int i = 0; i < count; ++i) {
+        DDS::MemberId id = data->get_member_id_at_index(i);
+        if (id == OpenDDS::XTypes::MEMBER_ID_INVALID) {
+            std::cerr << "Failed to get MemberId for element at index " << i << std::endl;
+            continue;
+        }
+
+        switch (elem_tk) {
+        case OpenDDS::XTypes::TK_SEQUENCE:
+        case OpenDDS::XTypes::TK_ARRAY:
+        case OpenDDS::XTypes::TK_STRUCTURE:
+        case OpenDDS::XTypes::TK_UNION: {
+            DDS::DynamicData_var nested_data;
+            DDS::ReturnCode_t ret = data->get_complex_value(nested_data, id);
+            if (ret != DDS::RETCODE_OK) {
+                std::cerr << "ge_complex_value for element at index " << i << " failed" << std::endl;
+            } else {
+	        std::string scoped_elem_name = namePrefix + "[" + std::to_string(i) + "]";
+		parseData(nested_data, scoped_elem_name);
+            }
+            continue;
+        }
+        }
+
+        // DataRow for each element
+        DataRow* data_row = new DataRow;
+        data_row->type = typekind_to_tckind(elem_tk);
+        data_row->isOptional = false;
+	std::string scoped_elem_name = namePrefix + "[" + std::to_string(i) + "]";
+	data_row->name = scoped_elem_name.c_str();
+        data_row->isKey = false;
```

Comment:
Could this be a key? For example, a struct with an array of 3 IDL longs as its key.

Reply:
I looked at this again and it's a bit more complicated than I originally thought. Setting `isKey` inside `parseCollection` is not hard -- we can pass the `is_key` value of the `MemberDescriptor` for this collection from the containing type. For setting `isKey` in `parseAggregated`, it needs to handle the implicit key case. But I'm also not sure if the `isKey` in `DataRow` has the same interpretation for keys as the KeyOnly/NestedKeyOnly in the core OpenDDS lib. Is this okay if I add TODOs for these.